### PR TITLE
[Comgr][hotswap] Append trampoline pool at a fresh vaddr instead of shifting post-.text addresses/symbols

### PR DESCRIPTION
**Pull request by xintin**
Fixes baked ISA/DWARF ref corruption; ROCm/llvm-project#3240 tests pass.

## Problem

With kernel-entry trampolines enabled on gfx1250 (`AMD_COMGR_HOTSWAP_ENTRY_TRAMPOLINES=1`), a kernel that touches a `__managed__` or `__device__` global reads and writes the wrong address, silently, with no crash or diagnostic. `Unit_hipModuleGetGlobal_Functional` reads 10 instead of 11 for a kernel like `__managed__ int x = 10; __global__ void GPU_func() { x++; }`. 
This is the corruption that `ROCm/llvm-project#3240` added regression tests for (both the ISA-literal and the DWARF variants).

## Root cause

`ElfView::growWithTrampolines` appended the trampoline pool by growing `.text` in place and shifting everything after it, so every following section `sh_addr`, every segment `p_vaddr`, and every post-`.text` symbol `st_value` was bumped by the pool size. But a fully-linked AMDGPU code object carries no relocations: the linker has already baked final PC-relative distances into the instruction stream, so a global access compiles to `s_get_pc_i64 s[0:1]` followed by `s_add_nc_u64 s[0:1], s[0:1], lit64(delta)` where `delta` is a baked literal with no relocation. The rewrite moved the symbol but not the baked literal (nor the absolute addresses in DWARF `.debug_*`), so the ISA and the debugger resolved the reference to the old location while the symbol table and `hipModuleGetGlobal` reported the new one, and the kernel dereferenced a stale slot.

## Approach

Append the pool at a fresh, page-aligned virtual address above every existing allocatable section, and leave everything else exactly in place, so all baked absolute and PC-relative addresses (and DWARF) stay valid because nothing existing moves.

- A new `ElfView::trampolinePoolVAddr()` is the single source of truth for the pool base.
- The pool is emitted as a new `PT_LOAD` (R+X) segment and an `SHF_ALLOC|SHF_EXECINSTR` section, and the program-header and section-header tables are relocated to end-of-file (they are metadata addressed via the ELF header, so relocating them corrupts nothing); there is no `.text` growth and no shifting, so the old `adjustSectionHeaders`, `adjustProgramHeaders`, and `adjustSymbolValues` are removed.
- Because the pool now sits far from `.text`, B0-to-A0 trampoline branch edges route through the existing `s_add_pc_i64` long branch; the branch and stub math is threaded through `PoolBaseOffset` (poolVAddr minus textAddr) in `emitToTrampoline` and `fixupTrampolineBranches`, and through the pool vaddr in the entry-stub delta and kernel-descriptor entry offset. Kernel-entry stubs already materialize a full 64-bit delta, so they reach from anywhere.
- Entry-stub idempotency reads the stub via a new `ElfView::dataAtVAddr` (which locates it in the pool section, not just `.text`), so a repeated rewrite stays a no-op.

## Tested

- The #3240 regression tests now pass (they fail on the pre-fix code): `GrowWithTrampolinesKeepsIsaReferenceConsistentWithSymbol` and `GrowWithTrampolinesKeepsDwarfConsistentWithSymbol`. The old `ShiftsAllocSectionSymbols` test was repurposed to `KeepsAllocSectionSymbols` (pins the no-shift behavior); the entry-stub MC tests and the wmma-split lit assertion were updated for the new layout.
- `HotswapMCTests` 44/44, `HotswapElfTests` 16/16, hotswap lit 53/53 (B0-to-A0 trampolines, long-branch, WMMA-split, and kernel-entry-trampoline including idempotency). clang-format clean.
- Remaining validation needs gfx1250 hardware: the host suite covers ELF well-formedness and address consistency, but the end-to-end `Unit_hipModuleGetGlobal_Functional` run (kernel reads 11, and the HSA loader maps the new `PT_LOAD` and relocated tables) should be confirmed on a gfx1250 device.

### CODE REVIEW

File: amd/comgr/src/comgr-hotswap-elf.cpp [hunk 1]
```diff
@@ -895,248 +880,56 @@ void ElfView::updateKernelDescriptor(StringRef KernelName, unsigned ExtraVgprs,
...
-      std::optional<uint64_t> Value =
-          checkedAddUint64(Sym.st_value, TrampTotal, "post-.text symbol value");
-      if (!Value)
-        return false;
-      uint64_t Value64 = *Value;
-      std::memcpy(Elf + SymOffset + offsetof(ELFT::Sym, st_value), &Value64,
-                  sizeof(Value64));
-    }
-    ++SectionIndex;
+    MaxAllocEnd = std::max(MaxAllocEnd, Shdr.sh_addr + Shdr.sh_size);
```
Comment (chinmaydd) [inline amd/comgr/src/comgr-hotswap-elf.cpp]: should use checkedAddUint64 ?

**Comment by chinmaydd** [comment]
It probably makes sense to add tests from https://github.com/ROCm/llvm-project/pull/3240 to this PR ?

**Comment by xintin** [comment]
> It probably makes sense to add tests from #3240 to this PR ?

it is included

**Comment by chinmaydd** [comment]
I'm not really a fan of the verbose comments but we can possibly clean those up later.

**Review by chinmaydd — APPROVED** [review]
There's a lot of fine-grained ELF surgery involved here and I'm sure we are missing some real corner cases but the passing tests provide confidence.

**Comment by xintin** [comment]
> There's a lot of fine-grained ELF surgery involved here and I'm sure we are missing some real corner cases but the passing tests provide confidence.

At this point it is based on the edge test cases provided and passing existing tests. Unfortunately/fortunately that is the best that can be done.
